### ISSUE TITLE
Additional streaming parameter types

### ISSUE THREAD

**Issue by DylanDmitri** [aspnet/SignalR#2594]
Currently the user can pass in `ChannelReader<T>`.

Want to expand this to accept the `Stream` class, possible the new `Pipeline` class, etc. Make a common pattern for adding on additional types, maybe somehow tied to `IEnumerableAsync`.

**Comment by anurse** [comment]
Nudging the cost up on this one as well :)

**Comment by DylanDmitri** [comment]
For the typescript client, there's no channels, so I ended up making a custom "streaming parameter" object.

This object has no async magic, no loops. It's just when you call "stream.write(5)" it generates and sends a stream item. The bad part is that you can only call this after making the invocation, whereas with a channel you can load items and then call afterwards.

```
// channel style
var channel = Channel.CreateUnbounded<Foo>();
foreach (foo in foos) {
  channel.Write(foo);
}
var result = await connection.invokeAsync("collectFoos", channel);
```

```
// flat placeholder style
var stream = connection.NewUploadStream();
var promise = connection.invoke("collectFoos", stream);
foreach (foo in foos) {
    await stream.write(foo);
}
var result = await promise;
```

The channel style one is definitely more elegant / more slick. The flat placeholder one better describes the underlying process; or the order in which things happen is more transparent.

What I want to do is just implement something similar in C#. Probably beef up the StreamPlaceholder class substantially, and have users call `stream.write(item)` like in typescript. This lets people do stuff like:

```
foreach (byte[] chunk in myFile) {
    await stream.write(chunk);
}
```

So basically this would allow for every iterable to be supported.

The downside is that there's something fun or magical about sending a channel "through to the server", and the nicer async staging results in cleaner code.

Thoughts?

**Comment by anurse** [comment]
My initial concern is that this is creating more ways to do the same things, and introducing new types that C# developers aren't familiar with. In general, we want to try to meet developers where they are, rather than teaching them new concepts. `Stream` and `Channel` are built-in to .NET, so developers will already understand them. Introducing a new concept means they have to learn something new just to use our framework.
